feat: Full and Final Settlement and Gratuity Fix

Anurag810 opened the pull request:
## What?

A checklist to keep a track of settling all the receivables, payables, assets allocated to an employee during employee exit

## Details

- Employee Details(section): Employee, Employee Name, Designation, Department
- Status: Paid / Unpaid
- Payables
- Receivables
- Totals
- Assets Allocated

## Workflow

- When an employee has resigned, set **Relieving Date** for the employee. FnF can only be created if relieving date is set.
- Select the employee in the FnF form
- The system will pull in some components in payables/receivables:
	- **Payables**: Salary Slip, Gratuity, Expense Claim, Bonus, Leave Encashment
	- **Receivables**: Loan, Employee Advance
- The default status for all will be **Unsettled**. Users can then select the respective document against each (like expense claims which are yet to be settled, bonus(additional salary), etc. and remove the components which do not require settling.
- Users can also add their own components or custom doctypes to this checklist.
- The amount and payable account will be fetched if applicable else it needs to be entered manually. 
	- For eg: 
		1. If a loan document is selected, then payment account is fetched and amount is set as `Total Payment - Total Amount Paid`
		2. If an Employee Advance is selected, then the payment account is set as advance's payment account and amount is set as `paid_amount - (claimed_amount + return_amount)`
		3. If an Expense Claim is selected, then the payment account is set as claim's payable account and amount is set as `grand_total - (total_amount_reimbursed + total_advance_amount)`
		4. For gratuity, payable account and amount are fetched as is.
		5. For Salary Slip, account is set as payroll payable account and amount is fetched from salary slip's `net pay`
-  Based on that total receivable/payable will be calculated 
- The system will automatically fetch assets allocated to the employee (eg: laptops, phones etc. allocated to the employee from the company). 
- These assets will have to be reallocated by the HR manager/user
- After this HR can mark all the components in Payables, Receivables, Assets as **Settled** after reviewing all the amounts and accounts.

![image](https://user-images.githubusercontent.com/24353136/151672775-5410a5a8-858f-452a-b065-70411c13fbe4.png)

![image](https://user-images.githubusercontent.com/24353136/151672781-81deafb3-76a4-40b7-8e0c-88753016b304.png)

## Settle with Journal Entry

- On submission, a button to create Journal Entry will be shown. Click on Create Journal Entry. This entry is made to record payment from the employee + payment done to the employee hence settling the account.
   
   ![image](https://user-images.githubusercontent.com/24353136/151672591-c2275f78-ee96-40b6-a25a-a39466de9268.png)

- This will map all your accounts and amounts to your journal entry. Select the final account from where the amount should get created if not already mapped. Save and Submit. FnF status should change to Paid.

   https://user-images.githubusercontent.com/24353136/151672810-d87433de-c50c-4a9d-b01e-382bed39042c.mp4

`no-docs` wiki does not support v14 yet

- [x] Development
- [x] Test

Anurag810 committed:
feat: Full and Final Settlement

Anurag810 committed:
removed option to pay via salary slip

Anurag810 committed:
feat: Create JV

Anurag810 committed:
test:fnf

Anurag810 committed:
fix: tracking asset movement

Anurag810 commented:
@ruchamahabal please start the review.

Anurag810 committed:
fix: sider and test

Anurag810 commented:
@ruchamahabal Test cases fixed and passed. Have a call with @asbasawaraj, will write documents when the branch for V14 Documents will be created.

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.json:
```diff
@@ -0,0 +1,204 @@
+   "fieldtype": "Column Break"
+  },
+  {
+   "fieldname": "payables",
```
Instead of labeling the child table, let's add a section heading since there is only one table in this section.
The same goes for Receivables and Assets Allocated.

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.json:
```diff
@@ -0,0 +1,204 @@
+   "label": "Employee Details"
+  },
+  {
+   "fieldname": "transaction_date",
```
Move Transaction Date and Status to the first section?

Anurag810 replied:
![image](https://user-images.githubusercontent.com/32095923/125764256-88443361-83dd-4ac6-86f8-1b9382095dfb.png)
10 fields in a single section don't sound good, I am moving that section above, section Employee Details for now.
I think this will work
![image](https://user-images.githubusercontent.com/32095923/125765210-8a40b628-325a-4542-82cc-1809daa7a053.png)

ruchamahabal replied:
The first section does not look good. You can rearrange the fields like this:

![image](https://user-images.githubusercontent.com/24353136/126268547-db8373b1-434f-4a00-b0e8-608c3a958ebe.png)

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.json:
```diff
@@ -0,0 +1,204 @@
+   "label": "Totals"
+  },
+  {
+   "fieldname": "total_payable_amount",
+   "fieldtype": "Currency",
+   "label": "Total Payable Amount"
```
This field should be `read_only` since it's auto calculated

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.json:
```diff
@@ -0,0 +1,204 @@
+   "fieldtype": "Column Break"
+  },
+  {
+   "fieldname": "total_receivable_amount",
+   "fieldtype": "Currency",
+   "label": "Total Receivable Amount,"
```
This field should be `read_only` since it's auto calculated

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+
+	def get_payable_component(self):
+		return [
+			"Salary Slip",
+			"Gratuity",
+			"Expense Claim",
+			"Bonus",
+			"Leave Encashment",
```
Since we know the reference doctypes for these components and receivable components, we should set the reference doctype too?

Anurag810 replied:
No, we cant, The component is only the name given It can be also like an Outstanding salary slip or advanced Salary slip It can be settled in other ways too.

ruchamahabal replied:
We can still set the reference doctype and users can change it if they want to, as in most cases the doctypes would be standard:

Salary Slip
Gratuity
Expense Claim
Additional Salary
Leave Encashment

Anurag810 replied:
Done

ruchamahabal commented on erpnext/accounts/doctype/journal_entry/journal_entry.py:
```diff
@@ -98,6 +100,16 @@ def update_advance_paid(self):
 				frappe.get_doc(voucher_type, voucher_no).set_total_advance_paid()
 
+	def update_status_for_full_and_final_statement(self):
+		fnf_docname = [d.reference_name for d in self.accounts if d.reference_type=="Full And Final Statement"]
+		if len(fnf_docname):
+			fnf_docname = fnf_docname[0]
+		if self.docstatus == 1:
+			frappe.db.set_value("Full And Final Statement", fnf_docname, "status", "Paid")
+		if self.docstatus == 2:
+			frappe.db.set_value("Full And Final Statement", fnf_docname, "status", "Unpaid")
```
You don't need to make a list and then choose the first element. You can directly iterate and update

```suggestion
		for entry in self.accounts:
			if entry.reference_type == "Full And Final Statement":
				if self.docstatus == 1:
					frappe.db.set_value("Full And Final Statement", entry.reference_name, "status", "Paid")
				elif self.docstatus == 2:
					frappe.db.set_value("Full And Final Statement", entry.reference_name, "status", "Unpaid")
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_assets/full_and_final_assets.json:
```diff
@@ -0,0 +1,63 @@
+   "label": "Description"
+  },
+  {
+   "fieldname": "asset_name",
+   "fieldtype": "Data",
+   "in_list_view": 1,
+   "label": "Asset Name",
+   "read_only": 1
+  },
+  {
+   "fieldname": "date",
+   "fieldtype": "Datetime",
+   "in_list_view": 1,
+   "label": "Date",
+   "read_only": 1
```
Asset Name and Date should be fetched from the selected Asset

![image](https://user-images.githubusercontent.com/24353136/125581842-f65a23e3-ce13-4f7e-ac92-e0f964c5dc21.png)

Anurag810 replied:
Sorry, it was supposed to be read-only, it will fetch Only pending asset movements towards employees. user should not be allowed to select.  it is handled through server script. so making table read only.

ruchamahabal commented on erpnext/hr/doctype/full_and_final_assets/full_and_final_assets.json:
```diff
@@ -0,0 +1,63 @@
+ "modified": "2021-07-03 04:24:10.526598",
+ "modified_by": "Administrator",
+ "module": "HR",
+ "name": "Full And Final Assets",
```
The child table name should not be plural. Only the table field in parent should be plural
```suggestion
 "name": "Full And Final Asset",
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_outstanding_statements/full_and_final_outstanding_statements.json:
```diff
@@ -0,0 +1,90 @@
+  },
+  {
+   "default": "0",
+   "fieldname": "considered_in_salary_slip",
+   "fieldtype": "Check",
+   "label": "Considered In Salary Slip"
```
```suggestion
   "fieldname": "payable_via_salary_slip",
   "fieldtype": "Check",
   "label": "Payable via Salary Slip"
```

Let's rename this to Payable via Salary Slip instead?
"Considered" seems ambiguous when the actual meaning is this component will be paid via journal entry if this is disabled.

Anurag810 replied:
Actually, It should not be like  "payable_via_salary_slip". Actually, some of the components like leave Encashment and Incentives are already Covered in the Salary slip This label Is for that. Open for Better Label suggestion, we can Have a discussion about that.

ruchamahabal replied:
How about **Paid via Salary Slip** then?

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_outstanding_statements/full_and_final_outstanding_statements.json:
```diff
@@ -0,0 +1,90 @@
+ "modified": "2021-07-02 15:57:54.254979",
+ "modified_by": "Administrator",
+ "module": "HR",
+ "name": "Full And Final Outstanding Statements",
```
```suggestion
 "name": "Full And Final Outstanding Statement",
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,93 @@
+			frm.add_custom_button(__('Create Journal Entry'), function () {
+				frm.events.create_journal_entry(frm);
+			});
+		}
```
The reference documents should have an Employee and Company filter
![image](https://user-images.githubusercontent.com/24353136/125578043-71fa76e1-1cf6-44d5-a7b6-87ee14c07490.png)

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,93 @@
+		frm.doc.total_receivable_amount = 0;
+
+		frm.doc.payables.forEach(element => {
+			frm.doc.total_payable_amount += element.amount;
```
Instead of updating the doc field in every iteration, use a variable to calculate the sum and then set the value with `flt`

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,93 @@
+		});
+
+		frm.doc.receivables.forEach(element => {
+			frm.doc.total_receivable_amount += element.amount;
```
Instead of updating the doc field in every iteration, use a variable to calculate the sum and then set the value with `flt`

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,93 @@
+		frm.doc.receivables.forEach(element => {
+			frm.doc.total_receivable_amount += element.amount;
+		});
+		cur_frm.refresh_field("total_payable_amount");
+		cur_frm.refresh_field("total_receivable_amount");
```
```suggestion
		frm.refresh_field("total_payable_amount");
		frm.refresh_field("total_receivable_amount");
```

Why `cur_frm`?

Anurag810 replied:
removed

ruchamahabal commented on erpnext/hr/doctype/full_and_final_outstanding_statements/full_and_final_outstanding_statements.json:
```diff
@@ -0,0 +1,90 @@
+   "fieldname": "status",
+   "fieldtype": "Select",
+   "label": "Status",
+   "options": "Settled\nUnsettled"
```
Let's also show status in grid view for every child table?

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,93 @@
+	},
+
+	refresh: function(frm) {
+		if (frm.doc.docstatus == 1 && frm.doc.status == "Unpaid") {
```
The individual components are not actually "Settled" unless we make a Journal Entry right? The doc cannot be submitted unless everything is settled and **create journal entry** option is shown after submission so how is this actually "Settled"?

Anurag810 replied:
After Creating Journal Entry The status of Docs will Change From Unpaid to Paid. HR needs to check the Documents, amount and have to mark it **settled** Before Creating the JV, It will Avoid **Wrong Payment** to employees if the HR User/Manager Mark It Stteled Manually. Open For better Workflow Suggestions.

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+		for data in self.payables:
+			if data.amount > 0 and not data.considered_in_salary_slip:
+				account_dict = {
+					"account": data.account,
+					"debit_in_account_currency": flt(data.amount, precision)
```
Can we also link reference type and name here?

Anurag810 replied:
No, we can't Because it will reveal the salary and amount specific to Employee to Accountant Maybe, it is not in every Company Policy. As we can Only provide Party and Party Type for Expense Claim And Employee Advance.

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+		for data in self.receivables:
+			if data.amount > 0:
+				account_dict = {
+					"account": data.account,
+					"credit_in_account_currency": flt(data.amount, precision)
```
Can we also link reference type and name here?

Anurag810 replied:
No, we can't Because it will reveal the salary and amount specific to Employee to Accountant Maybe, it is not in every Company Policy. As we can Only provide Party and Party Type for Expense Claim And Employee Advance.

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+@frappe.whitelist()
+def get_account_and_amount(ref_doctype, ref_document):
+	if ref_doctype == "Salary Slip" and ref_document:
+		payroll, amount = frappe.db.get_value("Salary Slip", ref_document, ["payroll_entry", "net_pay"])
+		payable_account = frappe.db.get_value("Payroll Entry", payroll, "payroll_payable_account")
```
What if the salary slip is not linked to any payroll entry yet? Can lead to unpacking errors

Anurag810 replied:
Fixed

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+
+@frappe.whitelist()
+def get_account_and_amount(ref_doctype, ref_document):
+	if ref_doctype == "Salary Slip" and ref_document:
```
Since you are checking the `ref_document` in every condition branch just add a condition at the very start of the function to return immediately if you don't find one.

Anurag810 replied:
Done

ruchamahabal replied:
Since you have added a check, in the beginning, you can remove the `and ref_document` conditions from everywhere in this function.

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+		else:
+			frappe.throw(_("Set Relieving Date for Employee: {0}").format(get_link_to_form("Employee", self.employee)))
+
+	def create_component_row(self, components, type):
```
```suggestion
	def create_component_row(self, components, component_type):
```
Not safe to use since `type` is a keyword in python?

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,170 @@
+			frappe.throw(_("Set Relieving Date for Employee: {0}").format(get_link_to_form("Employee", self.employee)))
+
+	def create_component_row(self, components, type):
+		data = {}
+		for component in components:
+			data["status"] = "Unsettled"
+			data["component"] = component
+			self.append(type, data)
```
unnecessary dict declaration?
```suggestion
		for component in components:
			self.append(component_type, {
				"status": "Unsettled",
				"component": component
			})
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/payroll/doctype/gratuity/gratuity.json:
```diff
@@ -49,26 +46,12 @@
    "reqd": 1
   },
-  {
-   "default": "1",
```
What happens to the existing data if we remove this field?

ruchamahabal reviewed (changes requested):
1. Doctype name can be Full **and** Final Statement / Full and Final Settlement instead of Full **And** Final Statement? A should not be capitalized in And.
2. Check for formatting issues. Try to keep uniform quote styles throughout 1 file.

Anurag810 committed:
fix: changes Requested

Anurag810 commented:
Failing Test "test_subcontracting_over_receipt" is not related to this PR

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,109 @@
+					istable: 0,
+					issingle: 0,
+					module: ["In", modules]
+				}
```
Add filter for employee and company here

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,109 @@
+});
+
+frappe.ui.form.on("Full and Final Outstanding Statement", {
+	// Not seems to be Working
+	// reference_document_type: function(frm, cdt, cdn) {
+	// 	let child = locals[cdt][cdn];
+	// 	console.log("Here")
+	// 	if (child.reference_document_type) {
+	// 		console.log("Here tooo")
+	// 		frm.set_query("reference_document", child.parentfield, function () {
+	// 			return {
+	// 				filters: {
+	// 					employee: frm.employee
+	// 				}
+	// 			};
+	// 		});
+	// 	}
+	// },
```
Remove this. Set filter in parent

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.js:
```diff
@@ -0,0 +1,109 @@
+		frm.doc.total_payable_amount = 0;
+		frm.doc.total_receivable_amount = 0;
+
+		frm.doc.payables.forEach(element => {
+			frm.doc.total_payable_amount += flt(element.amount);
+		});
+
+		frm.doc.receivables.forEach(element => {
+			frm.doc.total_receivable_amount += flt(element.amount);
```
Instead of updating the doc field in every iteration, use a variable to calculate the sum and then set the value with `flt`

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.json:
```diff
@@ -0,0 +1,201 @@
+   "fieldname": "status",
+   "fieldtype": "Select",
+   "label": "Status",
+   "options": "Paid\nUnpaid",
```
Add list view settings for the status

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/full_and_final_statement/full_and_final_statement.py:
```diff
@@ -0,0 +1,172 @@
+		return None
+
+	if ref_doctype == "Salary Slip" and ref_document:
+		payroll, amount = frappe.db.get_value("Salary Slip", ref_document, ["payroll_entry", "net_pay"])
+		payable_account = None
+		if payroll:
+			payable_account = frappe.db.get_value("Payroll Entry", payroll, "payroll_payable_account")
+			return [payable_account, amount]
```
```suggestion
		salary_details = frappe.db.get_value("Salary Slip", ref_document, ["payroll_entry", "net_pay"], as_dict=1)
		account = salary_details.net_pay
		payable_account = frappe.db.get_value("Payroll Entry", salary_details.payroll_entry, "payroll_payable_account") if salary_details.payroll_entry else None
		return [payable_account, amount]		
```
Can lead to unpacking errors if payroll entry isn't linked yet.

Anurag810 replied:
Done

Anurag810 committed:
fix: changes requested

Anurag810 committed:
fix: valication for Asset

Anurag810 commented:
@ruchamahabal

Anurag810 commented:
@ruchamahabal please merge

ruchamahabal committed:
Merge branch 'develop' into fnf

ruchamahabal committed:
fix: add filter for reference document only if relevant field is present

ruchamahabal committed:
fix: doctype cleanup

- add more fields to the list view and standard filter

- set title field

- incorrect field labels

ruchamahabal committed:
feat: add list view settings for FNF

ruchamahabal committed:
fix: incorrect reference type set in Journal Entry

ruchamahabal committed:
Merge branch 'develop' into fnf

ruchamahabal committed:
fix: validation message

ruchamahabal committed:
chore: add Full and Final Statement link to Workspace

ruchamahabal committed:
Merge branch 'develop' into fnf
